### ISSUE TITLE
Composite manifest support

### ISSUE THREAD

**Issue by rmarinvsn** [Orange-OpenSource/hasplayer.js#87]
Hello,

First of all, thank you for this amazing work you've done. Now my question, do you plan to add support for Smooth Streaming composite manifests? If you don't, could you give us some orientation to know what should we change (of course, all changes would be contributions to the project).

**Comment by bbert** [comment]
Hi,
What kind of scenario would you like to support with composite manifests?

**Comment by rmarinvsn** [comment]
HI bbert,

Thanks for the swift answer. Let me explain it. 
In our use case, our customers select cuts of various clips (can also be the same clip) with different mark in and mark out and can see the cuts all together.The number of cuts can vary by client, averaging 20/50 cuts but in some cases reaching 100 cuts.

Thanks again for all.
Best regards

**Comment by bbert** [comment]
Does it correspond to the "period" notion in MPEG-DASH?
Can you illustrate with a sample manifest?

**Comment by rmarinvsn** [comment]
Hi bbert

Regarding your first question, I think that a period in DASH is used to declare one cut, so I think that it does correspond. 

This is a very simple sample of the composite manifest in Smooth Streaming (1 cut of 1 clip). 
<?xml version="1.0"?>
<SmoothStreamingMedia
  MajorVersion="2"
  MinorVersion="1"
  Duration="38000000">
  <Clip Url="https://testingserver2/storage/Downloads/Primary/With You_5039620_22805/butterfly12345678912.ism/Manifest"  ClipBegin="15600000" ClipEnd="53600000">
<StreamIndex  Type="video"  Url="QualityLevels({bitrate})/Fragments(video={start time})"  Name="video"  Chunks="3"  QualityLevels="1">
  <QualityLevel    Index="0"    Bitrate="100000"    MaxWidth="640"    MaxHeight="360"    FourCC="AVC1" CodecPrivateData="000000016764001eac2ca40280bfe5c052828282a0000003002000000653b42c5c900000000168ebcd4940" />
<c  t="0"  d="20000000" />
<c  d="20000000" />
<c  d="20000000" />
</StreamIndex>
<StreamIndex  Type="audio"  Name="Chs 1+2 stereo"  Url="QualityLevels({bitrate})/Fragments(Chs 1+2 stereo={start time})"  Language="und"  Chunks="2"  QualityLevels="1">
  <QualityLevel    Index="0"    Bitrate="160000"    AudioTag="255"    SamplingRate="48000"    Channels="2"    BitsPerSample="16"    PacketSize="4"    CodecPrivateData="" />
<c  t="0"  d="50133333" />
<c  d="50133333" />
</StreamIndex>
</Clip>
</SmoothStreamingMedia>

Thanks again
Best Regards

**Comment by rmarinvsn** [comment]
`<?xml version="1.0"?>
<SmoothStreamingMedia
  MajorVersion="2"
  MinorVersion="1"
  Duration="38000000">
  <Clip Url="https://testingserver2/storage/Downloads/Primary/With You_5039620_22805/butterfly12345678912.ism/Manifest"  ClipBegin="15600000" ClipEnd="53600000">
<StreamIndex  Type="video"  Url="QualityLevels({bitrate})/Fragments(video={start time})"  Name="video"  Chunks="3"  QualityLevels="1">
  <QualityLevel    Index="0"    Bitrate="100000"    MaxWidth="640"    MaxHeight="360"    FourCC="AVC1" CodecPrivateData="000000016764001eac2ca40280bfe5c052828282a0000003002000000653b42c5c900000000168ebcd4940" />
<c  t="0"  d="20000000" />
<c  d="20000000" />
<c  d="20000000" />
</StreamIndex>
<StreamIndex  Type="audio"  Name="Chs 1+2 stereo"  Url="QualityLevels({bitrate})/Fragments(Chs 1+2 stereo={start time})"  Language="und"  Chunks="2"  QualityLevels="1">
  <QualityLevel    Index="0"    Bitrate="160000"    AudioTag="255"    SamplingRate="48000"    Channels="2"    BitsPerSample="16"    PacketSize="4"    CodecPrivateData="" />
<c  t="0"  d="50133333" />
<c  d="50133333" />
</StreamIndex>
</Clip>
</SmoothStreamingMedia>`

**Comment by bbert** [comment]
OK, then the idea would to add composite manifest parsing functionnality in MssParser.js so that you generate multiple periods, one for each clip.
Then StreamController is supposed to handle multiple periods and manage continuous playing of the successive periods, but we haven't really checked yet if it works fine.

**Comment by rmarinvsn** [comment]
Ok, thank you very much for this valuable information. We will try to do it and we will let you know the results.

Best Regards

**Comment by rmarinvsn** [comment]
Hello,

We've tried to modify the mss parser to support composite manifests. We have attached the mssparser.js file (please, remove the .jpg extension, we needed to add that extension to be able to upload the file). As a result, the code generates a multi-period mpd manifest, but the player is not able to play it back.

We have also attached a composite manifest just to show you how they look like. It points directly to the same samples that you have in your test player.

We've found also a multi-period mpd: 
[multiperiod mpd](http://dash.edgesuite.net/dash264/TestCases/5a/1/manifest.mpd)
and the [2.0 reference dash client ](http://dashif.org/reference/players/javascript/v2.0.0/samples/dash-if-reference-player/index.html) is able to play it, but your player doesn't. Does this mean that the current version (development branch) of your player doesn't support this feature?

We would highly appreciate if you could take a look to the modifications we have made to the mss parser. And we would like to know your opinion on the current status of the multi-period support.

Thank you very much for your time. Best regards

![mssparser js](https://cloud.githubusercontent.com/assets/13335832/13291181/0c5be038-db17-11e5-9101-78587e0f3a73.jpg)

![hotlisthasplayer zip](https://cloud.githubusercontent.com/assets/13335832/13290193/11881698-db13-11e5-8145-41d093be217e.jpg)

**Comment by bbert** [comment]
Hi,
First, I was unable to download your files, I get 403 error. However I managed to simulate manifest with multiple periods.
And I confirm it does not work at all and I do not know how it was initially supported by dash.js at the time we forked it.
We could try to implement multi-period support but on our side this is not at all a priority. Maybe you or any other contributor can try and we can provide some help or advice if needed on how to implement it.
Also, it would be good to see how it has been implemented in dash.js.
